plugins/woocommerce/src/Admin/Features/ProductBlockEditor/ProductTemplates/ProductTemplateRegistry.php

```diff
@@ -0,0 +1,79 @@
+<?php
+
+namespace Automattic\WooCommerce\Admin\Features\ProductBlockEditor\ProductTemplates;
+
+/**
+ * Product template registry.
+ */
+final class ProductTemplateRegistry {
+
+    /**
+	 * Class instance.
+	 *
+	 * @since 5.0.0
+	 * @var ProductTemplateRegistry|null
+	 */
+	private static $instance = null;
+
+    /**
+     * Templates.
+     */
+    protected $templates = array();
+
+    /**
+	 * Get the instance of the class.
+	 *
+	 * @return ProductTemplateRegistry The main instance.
+	 */
+	public static function get_instance() {
+		if ( null === self::$instance ) {
+			self::$instance = new self();
+		}
+
+		return self::$instance;
+	}
+
+    /**
+     * Register the default core product templates.
+     */
+    public function register_core_templates() {
+        $template_instances = array(
+            new SimpleProductTemplate(),
+            new VariableProductTemplate(),
+        );
+
+        foreach ( $template_instances as $instance ) {
+            $this->register( $instance->get_name(), $instance->get_template() );
+        }
+    }
+
+    /**
+     * Register a single template.
+     *
+     * @param string $name Name of the template
+     * @param array  $template Template layout.
+     */
+    public function register( $name, $template ) {
+        $this->templates[ $name ] = $template;
+    }
```

Comment:
Is it by design that a template can replace another template, if the same name is used? Who "wins"? Last one to register, given the POC implementation.

Maybe overriding/replacing a template deserves a more explicit call, such as `overrideRegisteredTemplate( $name, $template )`?

Reply:
Personally, I favor the ability to overwrite and think anyone who is using this is doing so with the intention to overwrite.  But I'm not opposed to creating an override function or locking this off completely with a warning once a template has been registered.

Reply:
Yeah, I suppose you are right that by intention you are calling this expecting your template to be registered with that name. I guess you really don't care of another template was already registered with that name. And if you do, we can provide a way to check if a template is already registered with a given name (`getTemplate( $name )`).